flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecHashJoin.java

```diff
@@ -0,0 +1,240 @@
+        this.rightAvgRowSize = rightAvgRowSize;
+        this.leftRowCount = leftRowCount;
+        this.rightRowCount = rightRowCount;
+        this.tryDistinctBuildRow = tryDistinctBuildRow;
+    }
+
+    @Override
+    @SuppressWarnings("Unchecked")
+    protected Transformation<RowData> translateToPlanInternal(PlannerBase planner) {
+        TableConfig config = planner.getTableConfig();
+
+        Transformation<RowData> lInput =
+                (Transformation<RowData>) getInputNodes().get(0).translateToPlan(planner);
+        Transformation<RowData> rInput =
+                (Transformation<RowData>) getInputNodes().get(1).translateToPlan(planner);
+
+        // get type
+        RowType lType = (RowType) getInputNodes().get(0).getOutputType();
+        RowType rType = (RowType) getInputNodes().get(1).getOutputType();
+
+        LogicalType[] keyFieldTypes =
+                IntStream.of(leftKeys).mapToObj(lType::getTypeAt).toArray(LogicalType[]::new);
+        RowType keyType = RowType.of(keyFieldTypes);
+
+        GeneratedJoinCondition condFunc =
+                JoinUtil.generateConditionFunction(config, nonEquiConditions, lType, rType);
+
+        // projection for equals
+        GeneratedProjection lProj =
+                ProjectionCodeGenerator.generateProjection(
+                        new CodeGeneratorContext(config),
+                        "HashJoinLeftProjection",
+                        lType,
+                        keyType,
+                        leftKeys);
+        GeneratedProjection rProj =
+                ProjectionCodeGenerator.generateProjection(
+                        new CodeGeneratorContext(config),
+                        "HashJoinRightProjection",
+                        rType,
+                        keyType,
+                        rightKeys);
+
+        Transformation<RowData> build;
+        Transformation<RowData> probe;
+        GeneratedProjection bProj;
+        GeneratedProjection pProj;
+        int[] bKeys;
+        int[] pKeys;
+        RowType bType;
+        RowType pType;
+        int bRowSize;
+        long bRowCount;
+        long pRowCount;
+        boolean reverseJoin = !leftIsBuild;
+        if (leftIsBuild) {
+            build = lInput;
+            bProj = lProj;
+            bType = lType;
+            bRowSize = leftAvgRowSize;
+            bRowCount = leftRowCount;
+            bKeys = leftKeys;
+
+            probe = rInput;
+            pProj = rProj;
+            pType = rType;
+            pRowCount = leftRowCount;
+            pKeys = rightKeys;
+        } else {
+            build = rInput;
+            bProj = rProj;
+            bType = rType;
+            bRowSize = rightAvgRowSize;
+            bRowCount = rightRowCount;
+            bKeys = rightKeys;
+
+            probe = lInput;
+            pProj = lProj;
+            pType = lType;
+            pRowCount = leftRowCount;
+            pKeys = leftKeys;
+        }
```

Comment:
give build/probe fields directly in the constructor ?

Reply:
the flag leftIsBuild cannot be reduced, so there is no need to specified build/probe side in the constructor, it is better to left them inside the ExecNode.